ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/QueryPlanningTool.java

```diff
@@ -0,0 +1,191 @@
+    private String defaultPrompt =
+        "You are an OpenSearch Query DSL generation assistant; try using the optional provided index mapping ${parameters.index_mapping:-}, specified fields ${parameters.query_fields:-}, and the given sample queries as examples, generate an OpenSearch Query DSL to retrieve the most relevant documents for the user provided natural language question: ${parameters.query_text}, please return the query dsl only in a string format, no other texts.\n";
+    @Getter
+    private Client client;
+    @Getter
+    private String modelId;
+    @Setter
+    @Getter
+    @VisibleForTesting
+    private Parser outputParser;
+    @Setter
+    @Getter
+    private String responseField;
+
+    public QueryPlanningTool(Client client, String modelId, String generationType, MLModelTool queryGenerationTool) {
+        this.client = client;
+        this.modelId = modelId;
+        this.generationType = generationType;
+        this.queryGenerationTool = queryGenerationTool;
+    }
+
+    @Override
+    public <T> void run(Map<String, String> parameters, ActionListener<T> listener) {
+        if (!parameters.containsKey(PROMPT_FIELD)) {
+            parameters.put(PROMPT_FIELD, defaultPrompt);
+        }
+        if (!validate(parameters)) {
+            listener.onFailure(new IllegalArgumentException("Empty parameters for QueryPlanningTool: " + parameters));
+            return;
+        }
+        ActionListener<T> modelListener = ActionListener.wrap(r -> {
+            try {
+                String queryString = (String) r;
+                if (queryString == null || queryString.isBlank() || queryString.isEmpty() || queryString.equals("null")) {
+                    StringSubstitutor substitutor = new StringSubstitutor(parameters, "${parameters.", "}");
+                    String defaultQueryString = substitutor.replace(this.defaultQuery);
+                    listener.onResponse((T) defaultQueryString);
+                } else {
+                    listener.onResponse((T) queryString);
+                }
+            } catch (Exception e) {
+                IllegalArgumentException parsingException = new IllegalArgumentException(
+                    "Error processing query string: " + r + ". Try using response_filter in agent registration if needed.",
+                    e
+                );
+                listener.onFailure(parsingException);
+            }
+        }, listener::onFailure);
+        queryGenerationTool.run(parameters, modelListener);
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    public String getVersion() {
+        return null;
+    }
+
+    @Override
+    public String getName() {
+        return this.name;
+    }
+
+    @Override
+    public void setName(String s) {
+        this.name = s;
+    }
+
+    @Override
+    public boolean validate(Map<String, String> parameters) {
```

Comment:
is there any critical fields to validate? Also this will never trigger as in line 76, we are doing `parameters.put(PROMPT_FIELD, defaultPrompt);`. So parameters shouldn't be empty?